sdks/common/runner-api/src/main/proto/beam_runner_api.proto

```diff
@@ -0,0 +1,436 @@
+  // may force runners to perform additional work prior to executing this ParDo)
+  //
+  // TODO: consider a DoFnPayload that holds these facts common to all DoFns
+  bool requires_window_access = 2;
+
+  // (Optional) A map from local input name to the specification for how to
+  // materialize and read it as a side input
+  map<string, SideInput> side_inputs = 3;
+
+  // (Optional) if the DoFn uses state, a map from id to the spec for the
+  // state cell.
+  map<string, StateSpec> state_specs = 4;
+
+  // (Optional) if the DoFn uses timers, a map from id to the spec for the
+  // timer.
+  map<string, TimerSpec> timer_specs = 5;
+}
+
+message StateSpec {
+  // TODO: AST for state spec
+}
+
+message TimerSpec {
+  // TODO: AST for timer spec
+}
+
+// The payload for the primitive Read transform.
+//
+// TODO: audit what needs to be in this payload
+message ReadPayload {
+
+  // (Required) The source for this Read
+  FunctionSpec source = 1;
+
+  // (Required) Whether the source is bounded or unbounded
+  bool is_bounded = 2;
+
+  // (Required) ...
+  bool does_not_require_splitting = 3;
+
+  // (Required) ...
+  bool produces_sorted_keys = 4;
+
+  // TODO: what else?
+}
+
+// The payload for the Window transform.
+//
+// TODO: consider making this non-primitive according to BEAM-<issue #>
+message WindowIntoPayload {
+  // (Required) the window fn for assigning windows
+  FunctionSpec window_fn = 1;
+}
+
+// The payload for the non-primitive Combine transform.
+message CombinePayload {
+
+  // (Required) The CombineFn for this Combine
+  FunctionSpec combine_fn = 1;
+
+  // (Required) A reference to the Coder to use for accumulators of the CombineFn
+  int64 accumulator_coder_ref = 2;
+
+  // (Optional) if the CombineFn uses side inputs, a map from the id to the spec
+  // for how to materialize and read the PCollection as a side input
+  map<string, SideInput> side_inputs = 3;
+}
+
+// A coder, the binary format for serialization and deserialization of data in
+// a pipeline.
+message Coder {
+
+  // (Required) A pipeline-scoped unique id for this node.
+  int64 id = 1;
+
+  // (Required) A cross-language, stable, unique identifier for the (possibly
+  // parametric) encoding.
+  string urn = 2;
+
+  // (Optional) If this coder is parametric, such as ListCoder(VarIntCoder),
+  // this is a list of the components. In order for encodings to be identical,
+  // the URN and all components must be identical, recursively.
+  repeated int64 component_coder_reference = 3;
+
+  // (Optional) An SDK-specific UDF implementing the coder.
+  //
+  // TODO: reconsider how to do this, aka CustomCoder
+  FunctionSpec custom_coder_fn = 4;
+}
+
+// A windowing strategy describes the window function, triggering, allowed
+// lateness, and accumulation mode for a PCollection.
+//
+// TODO: consider inlining field on PCollection if they aren't really a coherent whole
+message WindowingStrategy {
```

Comment:
It did until 5 minutes before I pushed. I could be convinced if there was a limitation?